schema/constraint_object.go

```diff
@@ -45,8 +49,72 @@ func (o Object) EmptyCompletionData(placeholder int, nestingLevel int) Completio
 }
 
 func (o Object) EmptyHoverData(nestingLevel int) *HoverData {
-	// TODO
-	return nil
+	if len(o.Attributes) == 0 {
+		return nil
+	}
+
+	attrNames := sortedObjectExprAttrNames(o.Attributes)
+
+	data := ""
+	if nestingLevel == 0 {
+		data += "```\n"
+	}
+
+	data += "{\n"
+	for _, name := range attrNames {
+		attr := o.Attributes[name]
+
+		cons, ok := attr.Constraint.(ConstraintWithHoverData)
+		if !ok {
+			return nil
+		}
+
+		hoverData := cons.EmptyHoverData(nestingLevel + 1)
+		if hoverData == nil {
+			return nil
+		}
+
+		attrFlags := []string{}
+		if attr.IsOptional {
+			attrFlags = append(attrFlags, "optional")
+		}
+		if attr.IsSensitive {
+			attrFlags = append(attrFlags, "sensitive")
+		}
+		attrComment := ""
+		if len(attrFlags) > 0 {
+			attrComment = fmt.Sprintf(" # %s", strings.Join(attrFlags, ", "))
+		}
+
+		data += fmt.Sprintf("%s%s = %s%s\n",
+			strings.Repeat("  ", nestingLevel+1),
+			name, hoverData.Content.Value, attrComment)
+	}
+	data += fmt.Sprintf("%s}", strings.Repeat("  ", nestingLevel))
+	if nestingLevel == 0 {
+		data += "\n```\n"
+	}
+
+	return &HoverData{
+		Content: lang.Markdown(data),
+	}
+}
+
+func sortedObjectExprAttrNames(attributes ObjectAttributes) []string {
+	if len(attributes) == 0 {
+		return []string{}
+	}
+
+	constraints := attributes
+	names := make([]string, len(constraints))
+	i := 0
+	for name := range constraints {
+		names[i] = name
+		i++
+	}
```

Comment:
```suggestion
	names := make([]string, 0, len(constraints))
	for name := range constraints {
		names = append(names, name)
	}
```

Nit: I think using `append` here makes it more readable.

Reply:
Even though the performance (memory) benefit is _probably_ minor, I'd say the readability gain is negligible too, so I'd prefer to keep this as is and benefit from lower memory consumption.

Reply:
Sure, we can keep this as is.

---

Yet, the memory consumptions for both approaches should be very similar. `make([]string, 0, len(constraints))` will initialize the slice with a zero-length and a capacity of `len(constraints)`, so Go internally already preallocates all the required elements. Therefore, adding up to `len(constraints)` elements to the slice means reusing the same backing array and no further allocations. The only difference is the repeated calls to the built-in `append` function, which has a small overhead compared to a direct assignment. Or am I missing something?